AgnostiqHQ/covalent
# graceful shutdown with cancellation

This is an update to an out-of-date PR #1258 which cancels pending or running workflows when the server is stopped.  This was not previously possible because of how we stopped the Uvicorn server using SIGKILL instead of SIGINT (gracefully).  When SIGINT is used instead, the second half of the FastAPI lifespan is able to execute before the server process terminates.

- [ ] I have added the tests to cover my changes.
- [x] I have updated the documentation and CHANGELOG accordingly.


To test this, consider the following workflow:

```
import covalent as ct
from time import sleep

@ct.electron
def task(x, y):
    sleep(10)
    return x + y

@ct.lattice
def workflow(x, y):
    for _ in range(10):
        x = task(x, y)
    return x
```

Execute this workflow, and then restart the server while it's running.  The workflow and task should be marked cancelled.
